### ISSUE TITLE
KTLS sending application data before Finished message

### ISSUE THREAD

**Issue by dcbeckett** [openssl/openssl#8896]
Commit: Fix error in BIO_get_ktls_send() and BIO_get_ktls_recv(), alters the BIO_get_ktls_send() and BIO_get_ktls_recv() functions definitions.

My understanding of the openssl codebase is limited, however after this commit and specifically the redefinition of these functions, the client and server skip the finished handshake payload and instead jump straight to sending tls application data

openssl s_server -accept 443
139799534412032:error:14094091:SSL routines:ssl3_read_bytes:data between ccs and finished:ssl/record/rec_layer_s3.c:1392:                                                                    
shutting down SSL
CONNECTION CLOSED

./openssl s_client -quiet -connect 10.0.0.2:443 -cipher 'ECDHE-RSA-AES128-GCM-SHA256' -tls1_2                                                         
139642010554624:error:140943F2:SSL routines:ssl3_read_bytes:sslv3 alert unexpected message:ssl/record/rec_layer_s3.c:1588:SSL alert number 10                                                
Using Kernel TLS for sending

@mattcaswell

**Comment by jayaram-intel** [comment]
Hi @dcbeckett ,

I want to work on this issue, please provide below information.

1. Which Linux Kernel version ( type `$ uname -r` on console) you are using?
2. Please provide OpenSSL code base up to specific commit ID or tag.

So that I will try to re-produce this issue on my side.

Thanks
Jayaram Matta

**Comment by jayaram-intel** [comment]
Hi @dcbeckett ,

I tried to reproduce the above issue with below environment setup
Ubuntu OS Version: **16.04**
Linux Kernel Version: **4.20.0-042000-generic**
OpenSSL code base: **Master branch** (code base up to 18th September 2019)
 
**I performed below steps:**
1.	Downloaded Linux kernel 4.20
2.	Compiled and installed the Linux kernel 4.20
3.	Enabled **kernel TLS** by using command `sudo modprobe tls` and verified it with `lsmod | grep tls`
4.	Cloned latest Openssl master code base ( till 18th September 2019) and executed below steps
•	./config enable-ktls
•	make
•	make test
•	sudo make install
5.	Executed s_server and s_client applications as mentioned below

**Terminal 1:** $ `sudo openssl s_server -accept 443`

> Using default temp DH parameters
> ACCEPT
> -----BEGIN SSL SESSION PARAMETERS-----
>/* some data */
> -----END SSL SESSION PARAMETERS-----
> Shared ciphers:ECDHE-RSA-AES128-GCM-SHA256
> Signature Algorithms: ECDSA+SHA256:ECDSA+SHA384:ECDSA+SHA512:Ed25519:Ed448:RSA-PSS+SHA256:RSA-PSS+SHA384:RSA-PSS+SHA512:RSA-PSS+SHA256:RSA-PSS+SHA384:RSA-PSS+SHA512:RSA+SHA256:RSA+SHA384:RSA+SHA512:ECDSA+SHA224:ECDSA+SHA1:RSA+SHA224:RSA+SHA1:DSA+SHA224:DSA+SHA1:DSA+SHA256:DSA+SHA384:DSA+SHA512
> Shared Signature Algorithms: ECDSA+SHA256:ECDSA+SHA384:ECDSA+SHA512:Ed25519:Ed448:RSA-PSS+SHA256:RSA-PSS+SHA384:RSA-PSS+SHA512:RSA-PSS+SHA256:RSA-PSS+SHA384:RSA-PSS+SHA512:RSA+SHA256:RSA+SHA384:RSA+SHA512:ECDSA+SHA224:ECDSA+SHA1:RSA+SHA224:RSA+SHA1:DSA+SHA224:DSA+SHA1:DSA+SHA256:DSA+SHA384:DSA+SHA512
> Supported Elliptic Curve Point Formats: uncompressed:ansiX962_compressed_prime:ansiX962_compressed_char2
> Supported Elliptic Groups: X25519:P-256:X448:P-521:P-384
> Shared Elliptic groups: X25519:P-256:X448:P-521:P-384
> CIPHER is ECDHE-RSA-AES128-GCM-SHA256
> Secure Renegotiation IS supported

**Terminal 2:** $ `openssl s_client -quiet -connect <ip>:443 -cipher 'ECDHE-RSA-AES128-GCM-SHA256' -tls1_2`
Able to send messages to server
Please let me know , If required I will share complete log.

**I didn’t see any ERROR messages as mentioned by you in your previous comment**
**Please try to use Linux kernel version 4.20 and verify from your end also.**
 

Thanks
Jayaram Matta

**Comment by morf** [comment]
Hi @jayaram-intel 

Error reproduced with:
```bash
# uname -r
4.20.17
```
and 70adc64632dde9359c8c1c23d01ef7f68d51382e commit

```bash
# openssl s_client -quiet -connect myhost:443 -cipher 'ECDHE-RSA-AES128-GCM-SHA256' -tls1_2
139822171342592:error:14094091:SSL routines:ssl3_read_bytes:data between ccs and finished:../ssl/record/rec_layer_s3.c:1340:
```

**Comment by jayaram-intel** [comment]
Hi @morf 

I have gone through KTLS related commits available in both OpenSSL Master branch and Linux Kernel (till 16th September 2019).

**Identified that KTLS related breaks are there after Linux kernel version 4.20.**

**I am able to Compile, Make test is passed and able to execute s_server and s_client applications with Linux kernel version 4.20.**

**Please try to use Linux kernel version 4.20 and verify from your end also.**

I think this issue is purely related to Linux kernel side KTLS breaks.

Thanks
Jayaram Matta

**Comment by morf** [comment]
@jayaram-intel 

You mean 4.20.00 kernel version? On 4.20.17 not work

**Comment by dcbeckett** [comment]
Openssl Commit 8450d0c784f8cec58e1b41c79fb3836b9f2acd5e
works fine for me on the 5.2 kernel. It's the following openssl commit which breaks it

Get Outlook for Android<https://aka.ms/ghei36>

________________________________

**Comment by jayaram-intel** [comment]
Hi @dcbeckett and @morf ,

Just now cross verified with **different Linux kernel versions**.
Able to compile and `make test` is passed for `Linux kernel versions 4.20 and 5.2.`
And `make test` is failed in `Linux kernel versions 5.0 and 5.2.11.`

Note: 
Linux Kernel TLS is enabled using `sudo modprobe tls`

Thanks
Jayaram Matta

**Comment by dcbeckett** [comment]
uname -r
5.2.14-arch1-1-ARCH

git clone https://github.com/openssl/openssl.git; cd openssl;

Step 1 - openssl commit that works - NOTE it should say "kernel tls" after the s_client command

git checkout 8450d0c784f8cec58e1b41c79fb3836b9f2acd5e; ./config enable-ktls ; make -j 16; export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:$(pwd); sudo modprobe tls;                                                                           
                                                                                 
cd apps; ./openssl s_client -connect google.com:443 -cipher 'ECDHE-RSA-AES128-GCM-SHA256' -tls1_2 -quiet -CApath /etc/ssl/certs/
                                                                                 
depth=1 C = US, O = Google Trust Services, CN = GTS CA 1O1                       
verify error:num=20:unable to get local issuer certificate                       
verify return:1                                                                  
depth=0 C = US, ST = California, L = Mountain View, O = Google LLC, CN = *.google.com
verify return:1                                                                  
Using Kernel TLS for sending                                                     
Using Kernel TLS for receiving                                                   
                        
Note: "Using kernel tls for sending"                                               
Now, repeat but with the next openssl repo commit                          
                                                                                 
cd ../; git checkout 3119ab3c9e6d211c461a245f3744893e17b6c193; make -j 16;   
    
cd apps; ./openssl s_client -connect google.com:443 -cipher 'ECDHE-RSA-AES128-GCM-SHA256' -tls1_2 -quiet -CApath /etc/ssl/certs/

depth=2 OU = GlobalSign Root CA - R2, O = GlobalSign, CN = GlobalSign            
verify return:1                                                                  
depth=1 C = US, O = Google Trust Services, CN = GTS CA 1O1                       
verify return:1                                                                  
depth=0 C = US, ST = California, L = Mountain View, O = Google LLC, CN = *.google.com
verify return:1                                                                  
139682726412608:error:14068044:SSL routines:do_ssl3_write:internal error:ssl/record/rec_layer_s3.c:807:
139682726412608:error:14068044:SSL routines:do_ssl3_write:internal error:ssl/record/rec_layer_s3.c:807:
Using Kernel TLS for sending                                                     


Slightly different error to what I posted above, but still a issue. I've been running openssl with ktls for the last 6 months with various kernels, we have fixed many of the ktls kernel bugs. I don't believe this is due to a kernel bug.
